plugins/azure-devops-backend/src/api/PullRequestsDashboardProvider.ts

```diff
@@ -0,0 +1,129 @@
+import {
+  DashboardPullRequest,
+  PullRequestOptions,
+  Team,
+  TeamMember,
+} from '@backstage/plugin-azure-devops-common';
+
+import { AzureDevOpsApi } from './AzureDevOpsApi';
+import { Logger } from 'winston';
+
+export class PullRequestsDashboardProvider {
+  private teams = new Map<string, Team>();
+
+  private teamMembers = new Map<string, TeamMember>();
+
+  private constructor(
+    private readonly logger: Logger,
+    private readonly azureDevOpsApi: AzureDevOpsApi,
+  ) {}
+
+  public static async create(
+    logger: Logger,
+    azureDevOpsApi: AzureDevOpsApi,
+  ): Promise<PullRequestsDashboardProvider> {
+    const provider = new PullRequestsDashboardProvider(logger, azureDevOpsApi);
+    await provider.readTeams();
+    return provider;
+  }
+
+  public async readTeams(): Promise<void> {
+    this.logger.info('Reading teams.');
+
+    let teams = await this.azureDevOpsApi.getAllTeams();
+
+    // This is used to filter out the default Azure Devops project teams.
+    teams = teams.filter(team =>
+      team.name && team.projectName
+        ? team.name !== `${team.projectName} Team`
+        : true,
+    );
+
+    this.teams = new Map<string, Team>();
+    this.teamMembers = new Map<string, TeamMember>();
+
+    await Promise.all(
+      teams.map(async team => {
+        const teamId = team.id;
+
+        if (teamId) {
+          const teamMembers = await this.azureDevOpsApi.getTeamMembers(team);
+
+          if (teamMembers) {
+            team.members = teamMembers.reduce((arr, teamMember) => {
+              const teamMemberId = teamMember.id;
+
+              if (teamMemberId) {
+                arr.push(teamMemberId);
+                const memberOf = [
+                  ...(this.teamMembers.get(teamMemberId)?.memberOf ?? []),
+                  teamId,
+                ];
+                this.teamMembers.set(teamMemberId, { ...teamMember, memberOf });
+              }
+
+              return arr;
+            }, [] as string[]);
+
+            this.teams.set(teamId, team);
+          }
+        }
+      }),
+    );
+  }
+
+  public async getDashboardPullRequests(
+    projectName: string,
+    options: PullRequestOptions,
+  ): Promise<DashboardPullRequest[]> {
+    const dashboardPullRequests =
+      await this.azureDevOpsApi.getDashboardPullRequests(projectName, options);
+
+    return dashboardPullRequests.map(pr => {
+      if (pr.createdBy?.id) {
+        const teamIds = this.teamMembers.get(pr.createdBy.id)?.memberOf;
+        pr.createdBy.teamIds = teamIds;
+        pr.createdBy.teamNames = teamIds?.map(
+          teamId => this.teams.get(teamId)?.name ?? '',
+        );
+      }
+
+      return pr;
+    });
+  }
+
+  public getUserTeamIds(email: string): string[] {
+    return (
+      this.getTeamMembers().find(teamMember => teamMember.uniqueName === email)
+        ?.memberOf ?? []
+    );
+  }
+
+  public async getAllTeams(): Promise<Team[]> {
+    if (!this.teams.size) {
+      await this.readTeams();
```

Comment:
So this basically happens once and then is cached forever. That may not be what you want to do, right? There should probably be an expiry, so that it refreshes every now and then.

And related but separate - it might make sense to store this in the form of `private cache: Promise<SomeCacheDataType> | undefined`. Anybody that wants access to some part of the cache just does `const cache = await this.getCache()`, and the private `getCache` method will know to either just return `this.cache` if it's non-undefined and hasn't expired, and otherwise assign a new promise to it and returning that (effectively making the caller wait for its completion automatically).

Reply:
Yeah it should have an expiry. I will come back to this in a follow up PR and look at using the cache pattern that you suggested.